base/src/main/scala/scalaz/data/Leibniz.scala

```diff
@@ -1,16 +1,216 @@
+    */
+  def coerce[L, H >: L, A >: L <: H, B >: L <: H]
+  (a: A, ab: Leibniz[L, H, A, B]): B =
+    ab.subst[λ[α => α]](a)
+
+  /**
+    * Equality is transitive relation and its witnesses can be composed
+    * in a chain much like functions.
+    *
+    * @see [[Leibniz.compose]]
+    * @see [[Leibniz.andThen]]
+    */
+  def trans[L, H >: L, A >: L <: H, B >: L <: H, C >: L <: H]
+  (bc: Leibniz[L, H, B, C], ab: Leibniz[L, H, A, B]): Leibniz[L, H, A, C] =
+    bc.subst[λ[`α >: L <: H` => Leibniz[L, H, A, α]]](ab)
+
+  /**
+    * Equality is symmetric and therefore can be flipped around.
+    *
+    * @see [[Leibniz.flip]]
+    */
+  def symm[L, H >: L, A >: L <: H, B >: L <: H]
+  (ab: Leibniz[L, H, A, B]): Leibniz[L, H, B, A] =
+    ab.subst[λ[`α >: L <: H` => Leibniz[L, H, α, A]]](refl)
+
+  /**
+    * Given `A === B` we can prove that `F[A] === F[B]`.
+    *
+    * @see [[lift2]]
+    * @see [[lift3]]
+    */
+  def lift[
+    L, H >: L, A >: L <: H, B >: L <: H,
+    LF, HF >: LF, F[_ >: L <: H] >: LF <: HF
+  ] (
+    eq: Leibniz[L, H, A, B]
+  ): Leibniz[LF, HF, F[A], F[B]] = {
+    type f[α >: L <: H] = Leibniz[LF, HF, F[A], F[α]]
+    eq.subst[f](refl_[LF, HF, F[A]])
+  }
+
+  /**
+    * Given `A === B` and `I === J` we can prove that `F[A, I] === F[B, J]`.
+    *
+    * @see [[lift]]
+    * @see [[lift3]]
+    */
+  def lift2[
+    L1, H1 >: L1, A1 >: L1 <: H1, B1 >: L1 <: H1,
+    L2, H2 >: L2, A2 >: L2 <: H2, B2 >: L2 <: H2,
+    LF, HF >: LF, F[_ >: L1 <: H1, _ >: L2 <: H2] >: LF <: HF
+  ] (
+      eq1: Leibniz[L1, H1, A1, B1],
+      eq2: Leibniz[L2, H2, A2, B2]
+  ): Leibniz[LF, HF, F[A1, A2], F[B1, B2]] = {
+    type f1[α >: L1 <: H1] = Leibniz[LF, HF, F[A1, A2], F[α, A2]]
+    type f2[α >: L2 <: H2] = Leibniz[LF, HF, F[A1, A2], F[B1, α]]
+    eq2.subst[f2](eq1.subst[f1](refl_[LF, HF, F[A1, A2]]))
+  }
+
+  /**
+    * Given `A === B`, `I === J`, and `M === N` we can prove that
+    * `F[A, I] === F[B, J]`.
+    *
+    * @see [[lift]]
+    * @see [[lift2]]
+    */
+  def lift3[
+    L1, H1 >: L1, A1 >: L1 <: H1, B1 >: L1 <: H1,
+    L2, H2 >: L2, A2 >: L2 <: H2, B2 >: L2 <: H2,
+    L3, H3 >: L3, A3 >: L3 <: H3, B3 >: L3 <: H3,
+    LF, HF >: LF, F[_ >: L1 <: H1, _ >: L2 <: H2, _ >: L3 <: H3] >: LF <: HF
+  ] (
+      eq1: Leibniz[L1, H1, A1, B1],
+      eq2: Leibniz[L2, H2, A2, B2],
+      eq3: Leibniz[L3, H3, A3, B3]
+  ): Leibniz[LF, HF, F[A1, A2, A3], F[B1, B2, B3]] = {
+    type f1[α >: L1 <: H1] = Leibniz[LF, HF, F[A1, A2, A3], F[α, A2, A3]]
+    type f2[α >: L2 <: H2] = Leibniz[LF, HF, F[A1, A2, A3], F[B1, α, A3]]
+    type f3[α >: L3 <: H3] = Leibniz[LF, HF, F[A1, A2, A3], F[B1, B2, α]]
+    eq3.subst[f3](eq2.subst[f2](eq1.subst[f1](refl_[LF, HF, F[A1, A2, A3]])))
+  }
+
+  /**
+    * Unsafe coercion between types. `unsafeForce` abuses `asInstanceOf` to
+    * explicitly coerce types. It is unsafe, but needed where Leibnizian
+    * equality isn't sufficient.
+    */
+  def unsafeForce[A, B]: A === B =
```

Comment:
Scalaz doesn't use WartRemover as far as I can tell, I will simply remove this annotation. All `asInstanceOf` require it, I was just being lazy with warnings in `leibniz` :wink: